### PR TITLE
WIP: Lease auth refactor

### PR DESCRIPTION
**Pull request by eschwartz**
## Proposed changes

- Refactor the /leases/{id}/auth controller to use new patterns for DB access and controller setup.
- Add a new `POST /leases/auth` to login to your active lease (based on your Cognito identity)
- Add functional test coverage for lease auth endpoints (had none, previously)

The new endpoint was trivial to add given the way the new code was structured, so it seemed an easy win to sneak in here.

I don't have any functional test coverage against the Cognito logic. I know @joshmarsh  is working on some functional test support for cognito integration, so I wanted to wait for that to get merged in before implementing the same pattern here.

-----

#### Manual testing

I deployed this PR to a sandbox environment, setup admin and user Cognito Users, and attempted to use the dce CLI to create and login to leases

**As a User by Lease ID**

Can only login to my own lease
```
~/dev/dce/modules$ dce leases list
[
	{
		"accountId": "<redacted>",
		"budgetAmount": 100,
		"budgetCurrency": "USD",
		"budgetNotificationEmails": [
			"a@b.com"
		],
		"createdOn": 1581610356,
		"expiresOn": 1582215156,
		"id": "392e14ab-44fe-4bc8-a813-14c84db58227",
		"lastModifiedOn": 1581610356,
		"leaseStatus": "Active",
		"leaseStatusModifiedOn": 1581610356,
		"leaseStatusReason": "Active",
		"principalId": "testuser"
	},
	{
		"accountId": "<redacted>",
		"budgetAmount": 100,
		"budgetCurrency": "USD",
		"budgetNotificationEmails": [
			"a@b.com"
		],
		"createdOn": 1581610793,
		"expiresOn": 1582215592,
		"id": "08182025-2dc6-4998-af82-8dba2784c463",
		"lastModifiedOn": 1581610793,
		"leaseStatus": "Active",
		"leaseStatusModifiedOn": 1581610793,
		"leaseStatusReason": "Active",
		"principalId": "testadmin"
	}
]

~/dev/dce/modules$ dce leases login -c 392e14ab-44fe-4bc8-a813-14c84db58227
<redacted>

~/dev/dce/modules$ dce leases login -c 08182025-2dc6-4998-af82-8dba2784c463
err:  [POST /leases/{id}/auth][401] postLeasesIdAuthUnauthorized
```

**As an admin user by Lease ID**

Can login to my own lease, and another user's lease

```
~/dev/dce/modules$ dce leases list
[
	{
		"accountId": "<redacted>",
		"budgetAmount": 100,
		"budgetCurrency": "USD",
		"budgetNotificationEmails": [
			"a@b.com"
		],
		"createdOn": 1581610356,
		"expiresOn": 1582215156,
		"id": "392e14ab-44fe-4bc8-a813-14c84db58227",
		"lastModifiedOn": 1581610356,
		"leaseStatus": "Active",
		"leaseStatusModifiedOn": 1581610356,
		"leaseStatusReason": "Active",
		"principalId": "testuser"
	},
	{
		"accountId": "<redacted>",
		"budgetAmount": 100,
		"budgetCurrency": "USD",
		"budgetNotificationEmails": [
			"a@b.com"
		],
		"createdOn": 1581610793,
		"expiresOn": 1582215592,
		"id": "08182025-2dc6-4998-af82-8dba2784c463",
		"lastModifiedOn": 1581610793,
		"leaseStatus": "Active",
		"leaseStatusModifiedOn": 1581610793,
		"leaseStatusReason": "Active",
		"principalId": "testadmin"
	}
]

~/dev/dce/modules$ dce leases login -c 392e14ab-44fe-4bc8-a813-14c84db58227
<redacted>

~/dev/dce/modules$ dce leases login -c 08182025-2dc6-4998-af82-8dba2784c463
<redacted>
```

**Login without Lease ID**

I put together a  branch of dce-cli that support `dce leases login` without a Lease ID param, so I can test the new `POST /leases/auth` endpoint
(I'll open a new PR for that in a minute)

```
~/dev/dce-cli$ ./dce leases login -p testuser
Adding credentials to .aws/credentials using AWS CLI

~/dev/dce-cli$ aws sts get-caller-identity --profile testuser
{
    "UserId": "AROAYM65MBBTL3FS6AN4M:testuser",
    "Account": "<redacted>",
    "Arn": "arn:aws:sts::<redacted>:assumed-role/DCEPrincipal-edan/testuser"
}
```

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Refactor (changes to code, which do not change application behavior)

## Checklist

- [x] I have filled out this PR template
- [x] I have read the [CONTRIBUTING](https://github.com/Optum/dce/blob/master/docs/CONTRIBUTING.md) doc
- [x] I have added automated tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (`README.md`, inline comments, etc.)
- [x] I have updated the `CHANGELOG.md` under a `## next` release, with a short summary of my changes

### CODE REVIEW

File: modules/variables.tf [hunk 1]
```diff
@@ -189,7 +189,7 @@ variable "cognito_identity_providers" {
 
 variable "cognito_roles_attribute_admin_name" {
   type    = string
-  default = ""
+  default = "Admin"
```
Comment (eschwartz) [inline modules/variables.tf]: Not sure how this was behaving before without a default value, but I'm thinking we need this here.
(this is how we map cognito `group` attributes to app-level roles.)

File: pkg/accountmanager/accountmanageriface/servicer.go [hunk 2]
```diff
@@ -5,14 +5,26 @@ package accountmanageriface
...
 // Servicer makes working with the Account Manager easier
+//go:generate mockery -name Servicer
 type Servicer interface {
 	// ValidateAccess creates a new Account instance
 	ValidateAccess(role *arn.ARN) error
+	// Retrieve credentials for the provided IAM Role ARN
+	Credentials(role *arn.ARN, roleSessionName string) Credentialer
+	// ConsoleURL generates a URL that may be used
+	// to login to the AWS web console for an account
+	ConsoleURL(creds Credentialer) (string, error)
```
Comment (eschwartz) [inline pkg/accountmanager/accountmanageriface/servicer.go]: Figured the AccountManager service was a good place for this abstraction, as it already has logic for managing assume-role creds

File: pkg/accountmanager/client.go [hunk 3]
```diff
@@ -11,48 +11,43 @@ import (
 	"github.com/aws/aws-sdk-go/aws/session"
 )
 
+//go:generate mockery -name clienter
 type clienter interface {
-	Config(roleArn *arn.ARN) *aws.Config
+	Config(roleArn *arn.ARN, roleSessionName string) *aws.Config
```
Comment (eschwartz) [inline pkg/accountmanager/client.go]: We have a business requirement for session names to match our principal IDs (it allows us to track CloudTrail events back to actual users)

File: pkg/accountmanager/client.go [hunk 4]
```diff
@@ -11,48 +11,43 @@ import (
...
 
 	// return no config for nil inputs
 	if roleArn == nil {
 		return nil
 	}
 
-	key := roleArn.String()
-
-	// check for cached config
-	if c.configs != nil && c.configs[key] != nil {
```
Comment (eschwartz) [inline pkg/accountmanager/client.go]: Not sure what's going on here, but this caching behavior seems to be causing the functional tests to fail, after their third run. I did try adding the new roleSessionName parameter to the cache key, but it didn't help.

I would run functional tests three times. In the third run (and all subsequent runs) the `POST /leases/auth` endpoint would return what _looked_ like valid credentials, but when you went to use them you'd get an _Invalid Credentials_ error from AWS.
I've been running functional tests over and over locally, and removing this caching logic seems to give me consistent success.

File: pkg/api/user.go [hunk 5]
```diff
@@ -29,8 +29,9 @@ type User struct {
 }
 
 // UserDetailer - used for mocking tests
+//go:generate mockery -name UserDetailer
 type UserDetailer interface {
-	GetUser(event *events.APIGatewayProxyRequest) *User
+	GetUser(reqCtx *events.APIGatewayProxyRequestContext) *User
```
Comment (eschwartz) [inline pkg/api/user.go]: We don't have direct access to the `APIGatewayProxyRequest` object in our controllers, using the new gorilla mux adapter.

File: pkg/accountmanager/client.go [hunk 6]
```diff
@@ -11,48 +11,43 @@ import (
...
-		return c.configs[key]
-	}
-
 	// new creds
-	creds := stscreds.NewCredentialsWithClient(c.sts, roleArn.String())
+	creds := stscreds.NewCredentialsWithClient(c.sts, roleArn.String(), func(p *stscreds.AssumeRoleProvider) {
```
Comment (kddejong) [inline pkg/accountmanager/client.go]: have you tested if the returned credentials have a duration you are planning on?

Comment (eschwartz) [reply pkg/accountmanager/client.go]: @kddejong -- I totally get what you're throwing at me now.

https://docs.aws.amazon.com/sdk-for-go/api/aws/credentials/stscreds/#NewCredentialsWithClient

>  The credentials will expire every 15 minutes and the role will be named after a nanosecond timestamp of this operation.

which is fine in golang world, where we're using the `Credentials.Retrieve()` method to auto-refresh as needed. But if we're handing those creds as static strings to end users, this is not going to work.

**Comment by eschwartz** [comment]
Another point from @kddejong, that I want to respond to

We should log the the generated Access Key ID, and the identity of who we delivered it to. That should help us out if we want to track who exactly used what set of credentials and when (eg. by inspecting CloudTrail logs)

**Comment by kddejong** [comment]
I do kind of wonder about a weird lock in if you don't need a lease ID.  If in the future someone wants to allow a user to have two leases this endpoint wouldn't work.  Maybe it seems like a feature of a client... to get their leases, see if there is one, and then log them in automatically if only one exists.

File: cmd/lambda/lease_auth/auth.go [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+	var lease *leases.Lease
+	var err error
+	if input.leaseID != nil {
+		// Grab the lease from the DB
+		lease, err = input.leaseService.Get(*input.leaseID)
+		if err != nil {
+			return nil, err
+		}
+	} else if user.Username != "" {
```
Comment (kddejong) [inline cmd/lambda/lease_auth/auth.go]: Is this a possible code path?  Either they have access to the lease or they don't?

File: cmd/lambda/lease_auth/auth.go [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+	// Other users can only login to their own leases
+	isLoginAllowed := user.Role == api.AdminGroupName ||
+		user.Username == *lease.PrincipalID
+	if !isLoginAllowed {
+		return nil, errors.NewUnauthorizedf("User %s does not have access to lease %s",
+			user.Username, *lease.ID)
+	}
+
+	// Lookup the Account, so we can get the Principal Role ARN
+	acct, err := input.accountService.Get(*lease.AccountID)
```
Comment (kddejong) [inline cmd/lambda/lease_auth/auth.go]: This isn't your work but this seems like a call we should try to remove.  Just store the principalRoleArn with the lease itself.

**Comment by eschwartz** [comment]
> I do kind of wonder about a weird lock in if you don't need a lease ID

@kddejong I'd be open to discussing that. Though in our particular case, I'm not sure if the dce-cli has access to our cognito username to figure this out.

File: cmd/lambda/lease_auth/auth.go (hunk 8 above)
Comment (eschwartz) [reply cmd/lambda/lease_auth/auth.go]: > Just store the principalRoleArn with the lease itself.

@kddejong I'm not sure I'm following. Are you suggesting a change to the Lease DB table?

File: cmd/lambda/lease_auth/auth.go (hunk 7 above)
Comment (eschwartz) [reply cmd/lambda/lease_auth/auth.go]: So logic here goes:

- if a Lease ID is provided, use that lease for auth'ing (line 81)
- If no lease is provided, but the request comes from an authenticated cognito user, lookup the active lease for that user (line 87)
- If we can't locate a lease using either of these approaches, return a 404 (line 101)

**Comment by nathanagood** [comment]
Per conversation with @eschwartz I'm closing this PR. May be revisted at some point.
